Firmware/lcd.cpp

```diff
@@ -809,151 +868,77 @@ void lcd_buttons_update(void)
-	0b10010,
-	0b01100,
-	0b00000,
-	0b00000,
-	0b00000,
-	0b00000};
-
-const uint8_t lcd_chardata_thermometer[8] PROGMEM = {
-	0b00100,
-	0b01010,
-	0b01010,
-	0b01010,
-	0b01010,
-	0b10001,
-	0b10001,
-	0b01110};
-
-const uint8_t lcd_chardata_uplevel[8] PROGMEM = {
-	0b00100,
-	0b01110,
-	0b11111,
-	0b00100,
-	0b11100,
-	0b00000,
-	0b00000,
-	0b00000}; //thanks joris
-
-const uint8_t lcd_chardata_refresh[8] PROGMEM = {
-	0b00000,
-	0b00110,
-	0b11001,
-	0b11000,
-	0b00011,
-	0b10011,
-	0b01100,
-	0b00000}; //thanks joris
-
-const uint8_t lcd_chardata_folder[8] PROGMEM = {
-	0b00000,
-	0b11100,
-	0b11111,
-	0b10001,
-	0b10001,
-	0b11111,
-	0b00000,
-	0b00000}; //thanks joris
-
-/*const uint8_t lcd_chardata_feedrate[8] PROGMEM = {
-	0b11100,
-	0b10000,
-	0b11000,
-	0b10111,
-	0b00101,
-	0b00110,
-	0b00101,
-	0b00000};*/ //thanks Sonny Mounicou
-
-/*const uint8_t lcd_chardata_feedrate[8] PROGMEM = {
-	0b11100,
-	0b10100,
-	0b11000,
-	0b10100,
-	0b00000,
-	0b00111,
-	0b00010,
-	0b00010};*/
-
-/*const uint8_t lcd_chardata_feedrate[8] PROGMEM = {
-	0b01100,
-	0b10011,
-	0b00000,
-	0b01100,
-	0b10011,
-	0b00000,
-	0b01100,
-	0b10011};*/
-
-const uint8_t lcd_chardata_feedrate[8] PROGMEM = {
-	0b00000,
-	0b00100,
-	0b10010,
-	0b01001,
-	0b10010,
-	0b00100,
-	0b00000,
-	0b00000};
-
-const uint8_t lcd_chardata_clock[8] PROGMEM = {
-	0b00000,
-	0b01110,
-	0b10011,
-	0b10101,
-	0b10001,
-	0b01110,
-	0b00000,
-	0b00000}; //thanks Sonny Mounicou
-
-void lcd_set_custom_characters(void)
-{
-	lcd_createChar_P(LCD_STR_BEDTEMP[0], lcd_chardata_bedTemp);
-	lcd_createChar_P(LCD_STR_DEGREE[0], lcd_chardata_degree);
-	lcd_createChar_P(LCD_STR_THERMOMETER[0], lcd_chardata_thermometer);
-	lcd_createChar_P(LCD_STR_UPLEVEL[0], lcd_chardata_uplevel);
-	lcd_createChar_P(LCD_STR_REFRESH[0], lcd_chardata_refresh);
-	lcd_createChar_P(LCD_STR_FOLDER[0], lcd_chardata_folder);
-	lcd_createChar_P(LCD_STR_FEEDRATE[0], lcd_chardata_feedrate);
-	lcd_createChar_P(LCD_STR_CLOCK[0], lcd_chardata_clock);
-}
-
-const uint8_t lcd_chardata_arr2down[8] PROGMEM = {
-	0b00000,
-	0b00000,
-	0b10001,
-	0b01010,
-	0b00100,
-	0b10001,
-	0b01010,
-	0b00100};
-
-const uint8_t lcd_chardata_confirm[8] PROGMEM = {
-	0b00000,
-	0b00001,
-	0b00011,
-	0b10110,
-	0b11100,
-	0b01000,
-	0b00000,
-	0b00000};
-
-void lcd_set_custom_characters_nextpage(void)
-{
-	lcd_createChar_P(LCD_STR_ARROW_2_DOWN[0], lcd_chardata_arr2down);
-	lcd_createChar_P(LCD_STR_CONFIRM[0], lcd_chardata_confirm);
+// #define DEBUG_CUSTOM_CHARACTERS
+
+static void lcd_print_custom(uint8_t c) {
+	uint8_t charToSend = pgm_read_byte(&Font[c - 0x80].alternate); // in case no empty slot is found, use the alternate character.
+	int8_t slotToUse = -1;
+
+	for (uint8_t i = 0; i < 8; i++) {
+		// first check if we already have the character in the lcd memory
+		if ((lcd_custom_characters[i] & 0x7F) == (c & 0x7F)) {
+			lcd_custom_characters[i] = c; // mark the custom character as used
+			charToSend = i; // send the found custom character id
```

Comment:
Would it make sense to add a constant or preprocessor symbol for `0x80` and `0x7F`?

This is my understanding:

* `0x7F` means custom character is unused
* `0x80` not sure what this one means

Reply:
Custom characters start at 0x80. 0x7F is the mask that is used for the custom character ID. Bit 7 has a different purpose. It is used to keep track of custom character usage. So if it's used on screen in the current frame, then bit 7 will be set. Frame start clears this bit for all characters. This bit is used to determine whether a character slot is available to be used (requires loading new character data to CGRAM, so slower) or potentially available (data already loaded, but it wasn't used yet on the current frame, will be discarded on the next frame start if unused this frame). A discarded character slot (or "empty" slot) is marked with the magic value 0x7F, not to be confused with the mask 0x7F used elsewhere.
I know this code is a bit messy because of these constants, but I really wanted to use every single bit of ram that I am allocating for this feature :)

Reply:
Hmmm, maybe I can use a bitset for this instead of masking. Not sure if the code will end up as efficient tho...

Reply:
> A discarded character slot (or "empty" slot) is marked with the magic value 0x7F, not to be confused with the mask 0x7F used elsewhere.

Ok this is good to know :) I just see the same value everywhere and thought they had the same "meaning"